detector service for breakage reporting

jdorweiler opened the pull request:
**Asana Task/Github Issue:** https://app.asana.com/1/137249556945/project/72649045549333/task/1211666118504301?focus=true

## Description

<!--
  Provide a terse summary of what the change is, detailed descriptions should belong in ship reviews or tech designs
-->

Building on #2048 with a more generalize framework for breakage reports.

**New detector service:**
- Registry + caching with configurable TTL
- Auto-registration from remote config during page load
- Optional auto-run with configurable delay (default 100ms)
- Domain gating + custom `shouldRun()` checks (only during auto-run)
- Multi-consumer design (any feature can import and use)

**Changed:**
- Architecture: Service with dynamic registration
- Caching: Built-in with TTL vs re-scan every call
- Configuration: Remote config vs function parameters
- Lifecycle: Auto-registers on page load + optional auto-run vs on-demand only
- Gating: Domain restrictions + custom checks
- Consumers: Any feature can use service

See the [injected/src/detectors/README.md](./injected/src/detectors/README.md) for more details 

## Testing Steps

- <!-- Include simple steps on how to check this change is working. Write "N/A" if not applicable. -->

## Checklist

<!--
  These questions are a friendly reminder to shipping code, if you're uncertain ask the AoR owners.
  It's also totally appropriate to not check some of these boxes, if they don't apply to your change.
-->
*Please tick all that apply:*

- [ ] I have tested this change locally
- [ ] I have tested this change locally in all supported browsers
- [ ] This change will be visible to users
- [ ] I have added automated tests that cover this change
- [ ] I have ensured the change is gated by config
- [ ] This change was covered by a ship review
- [ ] This change was covered by a tech design
- [ ] Any dependent config has been merged



<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduce on-demand bot/fraud detectors and wire them into breakage reporting, enable new feature across platforms, and add integration tests with mock pages.
> 
> - **Detectors**:
>   - Add on-demand utilities `runBotDetection` and `runFraudDetection` with shared helpers in `injected/src/detectors/**` and fallback `default-config`.
>   - New feature wrapper `injected/src/features/web-interference-detection.js` exposing `detectInterference` via messaging.
> - **Breakage Reporting**:
>   - Enhance `injected/src/features/breakage-reporting.js` to optionally include `detectorData` using `webInterferenceDetection` settings.
> - **Feature Enablement**:
>   - Register `webInterferenceDetection` in `injected/src/features.js` and enable on `apple`, `android`, `windows`, `chrome`, and `chrome-mv3`.
> - **Tests & Pages**:
>   - Expand `injected/integration-test/breakage-reporting.spec.js` with cases for clean page, Cloudflare, reCAPTCHA, and PerimeterX.
>   - Add test config enabling detectors and mock pages under `injected/integration-test/test-pages/breakage-reporting/**`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 7d8619cdb2e752f07cba2da5f9cae4c1eb81485e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

madblex commented on injected/src/detectors/detections/detection-base.js:
Will we still be needing this?

jdorweiler commented on injected/src/detectors/detector-service.js:
```diff
@@ -0,0 +1,82 @@
+ * @param {boolean} [options._autoRun] - Internal flag indicating auto-run (gates apply)
+ * @returns {Promise<Record<string, any>>} Object mapping detector IDs to their data
+ */
+export async function getDetectorBatch(detectorIds, options = {}) {
```
Yeah good point. I'll go with `getDetectorsData`. I don't think we need web prefix as we're not following that naming scheme elsewhere (e.g., getJsPerformanceMetrics, getExpandedPerformanceMetrics)

madblex commented:
> > This looks great. I added some comments and a couple more here:
> > ```
> > 1. I'd like to see how this integrates with content features, especially broker protection, i.e. the equivalent of [this](https://github.com/duckduckgo/content-scope-scripts/pull/2048/files#r2513656558)
> > 
> > 2. I see we removed `WebInterferenceDetection` content feature. Is it because you don't think we'll need a standalone feature for this? i.e. the detectors will **only** be called from other features.
> > 
> > 3. I see we're going the functional route (from classes). Not feeling strongly (I prefer functional most of the time), I just believe the trade-off is that interfaces will be more difficult to enforce. I think classes provide a bit of a better structure, but this is opinionated.
> > ```
> 
> 1. Commented here https://github.com/duckduckgo/content-scope-scripts/pull/2048/files#r2514564110
> 2. Yeah, I changed it so detectors auto register via detector-init.js during the load phase, and features import the service directly. So detector initialization is handled for you. Then you just need to add the message handler to broker protection (see my comment linked in 1)
> 3. I'm not strongly attached to this approach. I wanted to see what functional approach looked like vs classes but I'm open to switching to classes if you think the structure is worth it.

1. 👍 
2. Fine for me. We can always add a dedicated feature later if needed
3. No need to do the extra work. This will likely work fine as well. Thanks.

jonathanKingston commented on injected/src/content-scope-features.js:
```diff
@@ -44,6 +45,14 @@ export function load(args) {
 
     const bundledFeatureNames = typeof importConfig.injectName === 'string' ? platformSupport[importConfig.injectName] : [];
 
+    // Initialize detectors early so they're available when features init
```
I don't think we should put it that deep into the framework. Can we not trigger it in 'load' and whomever triggers firsts kicks it off for both?

jonathanKingston commented on injected/src/detectors/detector-init.js:
```diff
@@ -0,0 +1,133 @@
+            if (!checkGates(options, domains, shouldRun)) {
+                return null;
+            }
+            return await refresh(options);
```
I don't think bot-detection or fraud-detection has refresh but this I believe exposes it as having one meaning we will preference refresh right?

In:
```
    const runner = registration.refresh ?? registration.getData;
```
We're preferencing this function which is a noop as I understand it.

jdorweiler commented:
@jonathanKingston Thanks for the review. I refactored this into a WebInterferenceDetection ContentFeature, got rid of the standalone service/init, and switched to using getFeatureSetting() conditionalChanges for config. Let me know what you think.
